cloud/openssh/config_test.go

```diff
@@ -69,143 +69,227 @@ func TestHostOrMatchRegex(t *testing.T) {
+		existingSSHConfig := []byte("Host example.com\n\tUser example\n\tPort 1234\n")
+		in := fstest.MapFS{
+			".ssh": &fstest.MapFile{Mode: fs.ModeDir | 0700},
+			".ssh/config": &fstest.MapFile{
+				Data: existingSSHConfig,
+				Mode: 0644,
+			},
+		}
+		// Temporarily change the desired ~/.ssh/config so it contains
+		// the initial contents of the input ~/.ssh/config.
+		originalWantConfig := want[".ssh/config"]
+		defer func() { want[".ssh/config"] = originalWantConfig }()
+		want[".ssh/config"] = &fstest.MapFile{
+			Data: append(
+				[]byte("Include \"$HOME/.config/devbox/ssh/config\"\n"),
+				existingSSHConfig...,
+			),
+			Mode: 0644,
+		}
+
+		workdir := fsToDir(t, in)
+		t.Setenv("HOME", workdir)
+		if err := SetupDevbox(); err != nil {
+			t.Error("got SetupDevbox() error:", err)
+		}
+		got := os.DirFS(workdir)
+		fsEqual(t, got, want)
+	})
+	t.Run("AlreadySetup", func(t *testing.T) {
+		in := want
+		workdir := fsToDir(t, in)
+		t.Setenv("HOME", workdir)
+		if err := SetupDevbox(); err != nil {
+			t.Error("got SetupDevbox() error:", err)
+		}
+		got := os.DirFS(workdir)
+		fsEqual(t, got, want)
+	})
 }
 
-func TestAddVMKey(t *testing.T) {
-	workdir := duplicateDir(t, "testdata/add-key/in")
-	t.Setenv("HOME", workdir)
+//go:embed testdata/test.vm.devbox-vms.internal.golden
+var goldenVMKey []byte
 
-	// Must match the filename and content of
-	// testdata/add-key/out/.config/devbox/ssh/keys/test.vm.devbox-vms.internal
+func TestAddVMKey(t *testing.T) {
 	host := "test.vm.devbox-vms.internal"
-	key := `-----BEGIN OPENSSH PRIVATE KEY-----
-b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
-QyNTUxOQAAACAFBvrkKm1Zrwbpp7DAmTWj7i+S+cjGhAwm++fELg1OpwAAAKjD4XIZw+Fy
-GQAAAAtzc2gtZWQyNTUxOQAAACAFBvrkKm1Zrwbpp7DAmTWj7i+S+cjGhAwm++fELg1Opw
-AAAEAcvFwROtvcGVsdSg73Y+znyO9F6LFRxhWa7UJdGcjGzwUG+uQqbVmvBumnsMCZNaPu
-L5L5yMaEDCb758QuDU6nAAAAH2djdXJ0aXNAR3JlZ3MtTWFjQm9vay1Qcm8ubG9jYWwBAg
-MEBQY=
------END OPENSSH PRIVATE KEY-----
-`
-	if err := AddVMKey(host, key); err != nil {
+	in := fstest.MapFS{}
+	want := fstest.MapFS{
+		".config":                 &fstest.MapFile{Mode: fs.ModeDir | 0755},
+		".config/devbox":          &fstest.MapFile{Mode: fs.ModeDir | 0755},
+		".config/devbox/ssh":      &fstest.MapFile{Mode: fs.ModeDir | 0700},
+		".config/devbox/ssh/keys": &fstest.MapFile{Mode: fs.ModeDir | 0700},
+
+		".config/devbox/ssh/keys/" + host: &fstest.MapFile{
+			Data: goldenVMKey,
+			Mode: 0600,
+		},
+	}
+
+	workdir := fsToDir(t, in)
+	t.Setenv("HOME", workdir)
+	if err := AddVMKey(host, string(goldenVMKey)); err != nil {
 		t.Error("got AddKey(host, key) error:", err)
 	}
-	dirsEqual(t, workdir, "testdata/add-key/out")
+	got := os.DirFS(workdir)
+	fsEqual(t, got, want)
 }
 
-func duplicateDir(t *testing.T, dir string) string {
-	srcFS := os.DirFS(dir)
-	dstDir := t.TempDir()
-	err := fs.WalkDir(srcFS, ".", func(path string, _ fs.DirEntry, err error) error {
+func fsEqual(t *testing.T, got, want fs.FS) {
```

Comment:
these utility functions are great. I'm wondering if we can move them to an `internal/testing/fstests` package so they can be reused elsewhere....

Reply:
Sure! We can definitely do that if there's another package that would benefit from them.